lightning/src/offers/invoice.rs

```diff
@@ -491,19 +494,48 @@ where
 
 impl UnsignedBolt12Invoice {
 	fn new(invreq_bytes: &[u8], contents: InvoiceContents) -> Self {
+		// TLV record ranges applicable to invreq_bytes.
+		const NON_EXPERIMENTAL_TYPES: core::ops::Range<u64> = 0..INVOICE_REQUEST_TYPES.end;
+		const EXPERIMENTAL_TYPES: core::ops::Range<u64> =
+			EXPERIMENTAL_OFFER_TYPES.start..EXPERIMENTAL_INVOICE_REQUEST_TYPES.end;
+
+		let (_, _, _, invoice_tlv_stream, _, _, experimental_invoice_tlv_stream) =
+			contents.as_tlv_stream();
+
+		let mut bytes = Vec::with_capacity(
+			invreq_bytes.len()
+				+ invoice_tlv_stream.serialized_length()
+				+ experimental_invoice_tlv_stream.serialized_length(),
```

Comment:
Probably dense here but I'm not following why we're including the experimental invoice{_request} bytes in the non-experimental length. Something like this would read more clearly to me:
```rust
		let mut bytes = Vec::with_capacity(
			TlvStream::new(invreq_bytes)
				.range(NON_EXPERIMENTAL_TYPES)
				.map(|record| record.serialized_length())
				.sum::<usize>()
			.saturating_add(invoice_tlv_stream.serialized_length())
		);

Reply:
When the invoice is signed, we want `bytes` to be the exact size it will ultimately be to avoid re-allocating. So the size needs to include experimental TLVs. That is because `experimental_bytes` is copied into `bytes` after the signature is added.

So it will be the size of the invoice request (including experimental offer and invreq TLVs) minus the invoice request signature plus the invoice signature (so a wash) plus all invoice TLVs.

Reply:
Got it, thanks for that explanation.

Reply:
> When the invoice is signed, we want `bytes` to be the exact size it will ultimately be to avoid re-allocating. So the size needs to include experimental TLVs. That is because `experimental_bytes` is copied into `bytes` after the signature is added.
> 
> So it will be the size of the invoice request (including experimental offer and invreq TLVs) minus the invoice request signature plus the invoice signature (so a wash) plus all invoice TLVs.

Sorry I know I already ACK'd but I actually think this is useful info that would be nice to document in a comment. Feel free to save for the follow-up though.

Reply:
Adding now... and actual realized we aren't accounting for the signature in `UnsignedStaticInvoice::new`. 😬

Reply:
Updated now. Same for `UnsignedInvoiceRequest::new`.